### PR TITLE
Bugfix/maddpg mad4pg

### PR DESCRIPTION
**Pull request by DriesSmit**
## What?
Fix two bugs #358 (end of episode masking) and #357 (incorrect argument assignments) in the MADDPG and MAD4PG systems.
## Why?
Bugs in MADDPG/MAD4PG.
## How?
Made changes to MADDPG/MAD4PG trainers.
## Extra
.

### CODE REVIEW

File: mava/systems/tf/mad4pg/training.py [hunk 1]
```diff
@@ -563,15 +563,13 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
+                    rewards=rewards[agent],
+                    discounts=discount * agent_discount,
                     bootstrap_n=self._bootstrap_n,
                     loss_fn=losses.categorical,
                 )
-                self.critic_losses[agent] = tf.reduce_mean(critic_loss, axis=0)
-
                 # Actor learning.
                 obs_agent_feed = target_obs_trans[agent]
                 # TODO (dries): Why is there an extra tuple?
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:575]: Not sure what is meant by this?

File: mava/systems/tf/mad4pg/training.py [hunk 2]
```diff
@@ -610,13 +608,27 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
+                critic_mask = (
+                    tf.reshape(
+                        agent_discount[:, self._bootstrap_n :], critic_loss.shape
+                    )
+                    > 0.0
+                )
+                # TODO (dries): Is multiplication maybe better here? As assignment
+                # might not work with tf.function?
+                policy_loss = policy_loss[policy_mask]
+                critic_loss = critic_loss[critic_mask]  # Not masked because
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:628]: Will the fix be in a different PR?

File: mava/systems/tf/mad4pg/training.py [hunk 3]
```diff
@@ -610,13 +608,27 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
-                self.policy_losses[agent] = tf.reduce_mean(policy_loss, axis=0)
+                # Multiply by discounts to not train on padded data.
+                policy_mask = tf.reshape(agent_discount, policy_loss.shape) > 0.0
+                critic_mask = (
+                    tf.reshape(
+                        agent_discount[:, self._bootstrap_n :], critic_loss.shape
+                    )
+                    > 0.0
+                )
+                # TODO (dries): Is multiplication maybe better here? As assignment
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:625]: Not sure which assignment you are talking about?

File: mava/systems/tf/maddpg/training.py [hunk 4]
```diff
@@ -1312,16 +1312,14 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
 
                 # Critic loss.
                 critic_loss = recurrent_n_step_critic_loss(
-                    q_values,
-                    rewards[agent],
-                    discount * agent_discount,
-                    target_q_values,
+                    q_values=q_values,
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/training.py:1315]: Not sure if this is where some of the bugs got in, but I guess in the future we should always prefer keyword arguments as input instead of positional ones. 😄

Comment (DriesSmit) [reply mava/systems/tf/maddpg/training.py:1315]: Agreed :)

File: mava/systems/tf/mad4pg/training.py (hunk 1 above)
Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:575]: It doesn't really matter. There is an extra [0] that is needed. But I removed the message as it does not really matter.

File: mava/systems/tf/mad4pg/training.py (hunk 3 above)
Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:625]: There are two ways of masking. On is multiplying with the mask and the other indexing by the mask. I am not sure which one is better as multiplying has some theoretical issues (but maybe not practically). Indexing is probably better. Just not sure if tf.functions works well with it.

Comment (arnupretorius) [reply mava/systems/tf/mad4pg/training.py:625]: I agree with @KaleabTessera that we must be consistent. @DriesSmit what are the theoretical issues with multiplication?

Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:625]: I updated the code with Claude's suggestion:) Thanks for pointing this out @arnupretorius  and @KaleabTessera  :raised_hands:

**Review by arnupretorius — DISMISSED** [review]
Thanks for changes @DriesSmit! 🙂  

@KaleabTessera , @RuanJohn we should benchmark this before putting it in. @DriesSmit can you please send them a hyperparameter config.

File: mava/components/tf/losses/sequence.py [hunk 5]
```diff
@@ -27,45 +26,64 @@ def recurrent_n_step_critic_loss(
...
     target_q_values: tf.Tensor,
     rewards: tf.Tensor,
-    discounts: tf.Tensor,
     bootstrap_n: int,
-    loss_fn: Union[Type[trfl.td_learning], Type[losses.categorical]],
+    discount: float,
+    end_of_episode: tf.Tensor,
 ) -> tf.Tensor:
 
+    if type(q_values) == DiscreteValuedDistribution:
```
Comment (arnupretorius) [inline mava/components/tf/losses/sequence.py:34]: Perhaps use `isinstance` here?

File: mava/components/tf/losses/sequence.py [hunk 6]
```diff
@@ -27,45 +26,64 @@ def recurrent_n_step_critic_loss(
...
-        q_t = target_q_values
-        q_t.cut_dimension(axis=1, start=bootstrap_n)
-    n_step_rewards = rewards[:, :-bootstrap_n]
-    n_step_discount = discounts[:, :-bootstrap_n]
-    for i in range(1, bootstrap_n + 1):
-        n_step_rewards += rewards[:, i : i + seq_len - bootstrap_n]
+    q_tm1 = q_values
+
+    # The last values that rolled over do not matter because a
+    # mask is applied to it (hopefully).
```
Comment (arnupretorius) [inline mava/components/tf/losses/sequence.py:56]: I don't think it is good for us to have something happen "hopefully" 😅  Where is this expected to be handled? And by whom?

File: mava/systems/tf/mad4pg/training.py [hunk 7]
```diff
@@ -661,19 +661,18 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
                 # Cast the additional discount to match
                 # the environment discount dtype.
                 agent_discount = discounts[agent]
-                discount = tf.cast(self._discount, dtype=agent_discount.dtype)
+                # discount = tf.cast(self._discount, dtype=agent_discount.dtype)
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:664]: Remove?

File: mava/systems/tf/mad4pg/training.py [hunk 8]
```diff
@@ -661,19 +661,18 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
-                    rewards[agent],
-                    discount * agent_discount,
+                    q_values=q_values,
+                    target_q_values=target_q_values,
+                    rewards=rewards[agent],
+                    discount=self._discount,
+                    end_of_episode=agent_discount,
                     bootstrap_n=self._bootstrap_n,
-                    loss_fn=losses.categorical,
+                    # loss_fn=losses.categorical,
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:674]: Remove?

File: mava/systems/tf/mad4pg/training.py [hunk 9]
```diff
@@ -661,19 +661,18 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
                 )
-                self.critic_losses[agent] = tf.reduce_mean(critic_loss, axis=0)
-
                 # Actor learning.
                 obs_agent_feed = target_obs_trans[agent]
                 # TODO (dries): Why is there an extra tuple?
```
Comment (arnupretorius) [inline mava/systems/tf/mad4pg/training.py:678]: Is this still relevant here? I think a similar todo existed somewhere else and was removed?

File: mava/components/tf/losses/sequence.py [hunk 10]
```diff
@@ -27,45 +26,64 @@ def recurrent_n_step_critic_loss(
...
     n_step_rewards, _ = combine_dim(n_step_rewards)
-    n_step_discount, _ = combine_dim(n_step_discount)
 
-    critic_loss = loss_fn(
-        q_tm1,
-        n_step_rewards,
-        n_step_discount,
-        q_t,
+    q_t, _ = combine_dim(q_t)
+    eoe_mask, _ = combine_dim(eoe_mask)
```
Comment (arnupretorius) [inline mava/components/tf/losses/sequence.py:75]: I know acme/trfl have these short cryptic names, but I'm wondering if we shouldn't perhaps opt for fuller names that are more descriptive. It took one or two seconds for me to get that `eoe` probably means `end-of-episode`. I think being more descriptive can improve code readability. What do you think @DriesSmit and @KaleabTessera ?

File: mava/systems/tf/maddpg/training.py [hunk 11]
```diff
@@ -327,7 +327,9 @@ def _target_policy_actions(self, next_obs: Dict[str, np.ndarray]) -> Any:
             actions[agent] = self._target_policy_networks[agent_key](next_observation)
         return actions
 
-    @tf.function
+    # @tf.function
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/training.py:330]: Remove?

Comment (arnupretorius) [reply mava/systems/tf/maddpg/training.py:330]: Or should it be added back as the print suggests?

File: mava/systems/tf/maddpg/training.py [hunk 12]
```diff
@@ -1334,20 +1334,19 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
                 # Cast the additional discount to match
                 # the environment discount dtype.
                 agent_discount = discounts[agent]
-                discount = tf.cast(self._discount, dtype=agent_discount.dtype)
+                # discount = tf.cast(self._discount, dtype=agent_discount.dtype)
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/training.py:1337]: Remove?

File: mava/systems/tf/maddpg/training.py [hunk 13]
```diff
@@ -1334,20 +1334,19 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
-                    discount * agent_discount,
-                    target_q_values,
+                    q_values=q_values,
+                    target_q_values=target_q_values,
+                    rewards=rewards[agent],
+                    discount=self._discount,
+                    end_of_episode=agent_discount,
                     bootstrap_n=self._bootstrap_n,
-                    loss_fn=trfl.td_learning,
+                    # loss_fn=trfl.td_learning,
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/training.py:1347]: Remove?

File: mava/systems/tf/maddpg/training.py [hunk 14]
```diff
@@ -1334,20 +1334,19 @@ def _forward(self, inputs: reverb.ReplaySample) -> None:
...
                 )
 
-                self.critic_losses[agent] = tf.reduce_mean(critic_loss, axis=0)
-
                 # Actor learning.
                 obs_agent_feed = target_obs_trans[agent]
                 # TODO (dries): Why is there an extra tuple?
```
Comment (arnupretorius) [inline mava/systems/tf/maddpg/training.py:1352]: Remove?

**Review by arnupretorius — COMMENTED** [review]
Thanks @DriesSmit!  🙂 

Just see my few comments.

File: mava/components/tf/losses/sequence.py (hunk 6 above)
Comment (DriesSmit) [reply mava/components/tf/losses/sequence.py:56]: This is handled already :) I will remove the hopefully.

File: mava/systems/tf/mad4pg/training.py (hunk 7 above)
Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:664]: Removed.

File: mava/systems/tf/mad4pg/training.py (hunk 8 above)
Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:674]: Removed.

File: mava/systems/tf/mad4pg/training.py (hunk 9 above)
Comment (DriesSmit) [reply mava/systems/tf/mad4pg/training.py:678]: Removed.

File: mava/systems/tf/maddpg/training.py (hunk 13 above)
Comment (DriesSmit) [reply mava/systems/tf/maddpg/training.py:1347]: Fixed.

File: mava/systems/tf/maddpg/training.py (hunk 12 above)
Comment (DriesSmit) [reply mava/systems/tf/maddpg/training.py:1337]: Fixed.

File: mava/components/tf/losses/sequence.py (hunk 10 above)
Comment (DriesSmit) [reply mava/components/tf/losses/sequence.py:75]: Agreed. I fixed it :)

**Review by arnupretorius — DISMISSED** [review]
Thanks @DriesSmit! Looks good to me 🙂

File: mava/components/tf/losses/sequence.py [hunk 15]
```diff
@@ -27,45 +28,71 @@ def recurrent_n_step_critic_loss(
...
-    critic_loss = loss_fn(
-        q_tm1,
-        n_step_rewards,
-        n_step_discount,
-        q_t,
-    )
+    # Create the end of episode mask.
+    ones_mask = tf.ones(shape=r_shape[:-1] + (r_shape[-1] - bootstrap_n,))
+    zeros_mask = tf.zeros(shape=r_shape[:-1] + (bootstrap_n,))
+    end_of_episode_mask, _ = combine_dim(tf.concat([ones_mask, zeros_mask], axis=1))
```
Comment (KaleabTessera) [inline mava/components/tf/losses/sequence.py:77]: What is purpose of this var? Seems like it always set to the same things (as opposed to reading it from replay). If this a flag to indicate end of episode, shouldn't this be read from replay?

**Review by KaleabTessera — COMMENTED** [review]
Thanks @DriesSmit , just a minor comment re `end_of_episode_mask` :+1:

File: mava/components/tf/losses/sequence.py [hunk 16]
```diff
@@ -27,45 +28,71 @@ def recurrent_n_step_critic_loss(
...
-    discounts: tf.Tensor,
     bootstrap_n: int,
-    loss_fn: Union[Type[trfl.td_learning], Type[losses.categorical]],
+    discount: float,
+    end_of_episode: tf.Tensor,
 ) -> tf.Tensor:
 
+    # d: discount * done
+    # bootstrap_n=1 is the normal return of Q_t-1 = R_t-1 + d * Q_t
+    # bootstrap_n=seq_len is the Q_t_1 = discounted sum of rewards return
```
[the remark this answers is not in the document]
Comment (DriesSmit) [reply mava/components/tf/losses/sequence.py:38]: Good catch :smile: I change it to a doc format. It is a note for anyone trying to understand what to set `bootstrap_n` to.

File: mava/components/tf/losses/sequence.py (hunk 15 above)
Comment (DriesSmit) [reply mava/components/tf/losses/sequence.py:77]: @KaleabTessera so we have two masks in the loss function. One is to mask zero-padded data and the other one (`end_of_episode_mask`) masks out the value function for where the data is not available for the target_value to bootstrap from. I am happy to have a call with you to further explain this. It would be nice if I can have a second pair of eyes on my code :slightly_smiling_face:
